lib/utils/prompt/context_reader_test.go

```diff
@@ -0,0 +1,211 @@
+	t.Cleanup(func() { pw.Close() })
+
+	write := func(t *testing.T, s string) {
+		_, err := pw.Write([]byte(s))
+		assert.NoError(t, err, "Write failed")
+	}
+
+	ctx := context.Background()
+	cr := NewContextReader(pr)
+
+	t.Run("simple read", func(t *testing.T) {
+		go write(t, "hello")
+		buf, err := cr.ReadContext(ctx)
+		require.NoError(t, err)
+		require.Equal(t, string(buf), "hello")
+	})
+
+	t.Run("reclaim abandoned read", func(t *testing.T) {
+		done := make(chan struct{})
+		cancelCtx, cancel := context.WithCancel(ctx)
+		go func() {
+			time.Sleep(1 * time.Millisecond) // give ReadContext time to block
+			cancel()
+			write(t, "after cancel")
+			close(done)
+		}()
+		buf, err := cr.ReadContext(cancelCtx)
+		require.ErrorIs(t, err, context.Canceled)
+		require.Empty(t, buf)
+
+		<-done // wait for write
+		buf, err = cr.ReadContext(ctx)
+		require.NoError(t, err)
+		require.Equal(t, string(buf), "after cancel")
+	})
+
+	t.Run("close ContextReader", func(t *testing.T) {
+		go func() {
+			time.Sleep(1 * time.Millisecond) // give ReadContext time to block
+			assert.NoError(t, cr.Close(), "Close errored")
+		}()
+		_, err := cr.ReadContext(ctx)
+		require.ErrorIs(t, err, ErrReaderClosed)
+
+		// Subsequent reads fail.
+		_, err = cr.ReadContext(ctx)
+		require.ErrorIs(t, err, ErrReaderClosed)
+
+		// Ongoing read after Close is dropped.
+		write(t, "unblock goroutine")
+		buf, err := cr.ReadContext(ctx)
+		assert.ErrorIs(t, err, ErrReaderClosed)
+		assert.Empty(t, buf, "buf not empty")
+
+		// Multiple closes are fine.
+		assert.NoError(t, cr.Close(), "2nd Close failed")
+	})
+
+	// Re-creating is safe because the tests above leave no "pending" reads.
+	cr = NewContextReader(pr)
+
+	t.Run("close underlying reader", func(t *testing.T) {
+		go func() {
+			write(t, "before close")
+			pw.CloseWithError(io.EOF)
+		}()
+
+		// Read the last chunk of data successfully.
+		buf, err := cr.ReadContext(ctx)
+		require.NoError(t, err)
+		require.Equal(t, string(buf), "before close")
+
+		// Next read fails because underlying reader is closed.
+		buf, err = cr.ReadContext(ctx)
+		require.ErrorIs(t, err, io.EOF)
+		require.Empty(t, buf)
+	})
+}
+
+func TestContextReader_ReadPassword(t *testing.T) {
+	resetTermAfterTests(t)
+
+	pr, pw := io.Pipe()
+	write := func(t *testing.T, s string) {
+		_, err := pw.Write([]byte(s))
+		assert.NoError(t, err, "Write failed")
+	}
+
+	restoreCalled := false
+	termIsTerminal = func(fd int) bool {
+		return true
+	}
+	termGetState = func(fd int) (*term.State, error) {
+		return &term.State{}, nil
+	}
+	termReadPassword = func(fd int) ([]byte, error) {
+		const bufLen = 1024 // arbitrary, big enough for our data
+		data := make([]byte, bufLen)
+		n, err := pr.Read(data)
+		data = data[:n]
+		return data, err
+	}
+	termRestore = func(fd int, oldState *term.State) error {
+		restoreCalled = true
+		return nil
+	}
+
+	cr := NewContextReader(pr)
+	cr.fd = 15 // arbitrary, doesn't matter because term functions are mocked.
```

Comment:
Could you open `os.DevNull` for readwrite and use that file's Fd for this, just in case?

Reply:
Sure, done.